### ISSUE TITLE
Support for proposed folding provider.

### ISSUE THREAD

**Issue by angelozerr** [eclipse/lsp4j#169]
VSCode provides folding provider for HTML, JSON.

VSCode has defined in their proposed https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.proposed.d.ts#L37

See https://github.com/Microsoft/language-server-protocol/issues/418

It should be cool if lsp4j could provide this feature. Thanks!

**Comment by angelozerr** [comment]
Is there any chance that someone could work on this issue?

**Comment by angelozerr** [comment]
Here the official spec https://microsoft.github.io/language-server-protocol/specification#textDocument_foldingRange

**Comment by spoenemann** [comment]
> Is there any chance that someone could work on this issue?

Maybe later this month. Contributions are welcome :-)

**Comment by angelozerr** [comment]
Thank a lot @spoenemann ! I will give you some feedback when I will implement folding support on client side (with LSP4e) and server side (with my XML Language Server lsp4xml)

**Comment by spoenemann** [comment]
[v0.5.0](https://github.com/eclipse/lsp4j/milestone/3) is on September 7th. We'll also have a milestone this week that includes it.

**Comment by angelozerr** [comment]
Thanks @spoenemann !

**Comment by angelozerr** [comment]
@spoenemann there were a problem in the folding specification which was fixed at https://github.com/Microsoft/language-server-protocol/commit/09ecb5ecd97cc0888f591cfe2065dce4c6208df4

In other words, you need to change to remove the 's' of foldingRange like this:

```java
CompletableFuture<List<FoldingRange>> foldingRange(FoldingRangeRequestParams params)
```

(foldingRanges to foldingRange)

**Comment by angelozerr** [comment]
Thanks @spoenemann !

**Comment by angelozerr** [comment]
@spoenemann just to give you feedback I have implemented folding on client side with LSP4E https://bugs.eclipse.org/bugs/show_bug.cgi?id=537706 and it works great (VSCode CSS Language Server folding is working).

For client side, I'm waiting for that 0.5.0 SNAPSHOT release will contain your fix with foldingRange to try it with my XML Language Server written in Java.

**Comment by spoenemann** [comment]
The current state is already available in the 0.5.0.M1 milestone release.

**Comment by angelozerr** [comment]
> The current state is already available in the 0.5.0.M1 milestone release.

0.5.0 SNAPSHOT is avaialbel today. I confirm you that my XML Language Server (server side) works well. Thanks again @spoenemann !
